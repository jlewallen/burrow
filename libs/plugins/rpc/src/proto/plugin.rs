use super::fsm::{Machine, Transition};
use super::*;
use super::{Payload, PayloadMessage, Query};
use anyhow::Result;
use tracing::warn;

const DEFAULT_DEPTH: u32 = 2;

type PluginTransition = Transition<PluginState, Query>;

#[derive(Debug, PartialEq, Eq)]
pub enum PluginState {
    Uninitialized,
    Initialized,
    Failed,
    Resolving,
}

type PluginMachine = Machine<PluginState>;

impl PluginMachine {
    fn new() -> Self {
        Self {
            state: PluginState::Uninitialized,
        }
    }
}

pub trait PluginResponses {
    fn surroundings(surroundings: &Surroundings) -> PluginTransition;
}

pub struct DefaultResponses {}

impl PluginResponses for DefaultResponses {
    fn surroundings(surroundings: &Surroundings) -> PluginTransition {
        let keys = match &surroundings {
            Surroundings::Living {
                world,
                living,
                area,
            } => vec![world, living, area],
        };

        let lookups = keys.into_iter().map(|k| LookupBy::Key(k.clone())).collect();
        let lookup = Query::Lookup(DEFAULT_DEPTH, lookups);
        PluginTransition::Send(lookup, PluginState::Resolving)
    }
}

#[derive(Debug)]
pub struct PluginProtocol<R>
where
    R: PluginResponses,
{
    session_key: Option<String>,
    machine: PluginMachine,
    _marker: std::marker::PhantomData<R>,
}

impl<R> PluginProtocol<R>
where
    R: PluginResponses,
{
    pub fn new() -> Self {
        Self {
            session_key: None,
            machine: PluginMachine::new(),
            _marker: Default::default(),
        }
    }

    #[cfg(test)]
    pub fn new_with_session_key(session_key: String) -> Self {
        Self {
            session_key: Some(session_key),
            machine: PluginMachine::new(),
            _marker: Default::default(),
        }
    }

    #[cfg(test)]
    pub fn session_key(&self) -> Option<&str> {
        self.session_key.as_deref()
    }

    pub fn message<B>(&self, body: B) -> Message<B> {
        Message {
            session_key: self.session_key.clone().expect("A session key is required"),
            body,
        }
    }

    pub fn apply(
        &mut self,
        message: &PayloadMessage,
        sender: &mut Sender<QueryMessage>,
    ) -> Result<()> {
        let transition = self.handle(message).map_message(|m| QueryMessage {
            session_key: self.session_key.as_ref().unwrap().clone(),
            body: Some(m),
        });

        self.machine.apply(transition, sender)?;

        Ok(())
    }

    fn handle(&mut self, message: &PayloadMessage) -> PluginTransition {
        match (&self.machine.state, &message.body) {
            (PluginState::Uninitialized, Payload::Initialize(session_key)) => {
                self.session_key = Some(session_key.to_owned());

                PluginTransition::Direct(PluginState::Initialized)
            }
            (PluginState::Initialized, Payload::Surroundings(surroundings)) => {
                R::surroundings(surroundings)
            }
            (PluginState::Resolving, Payload::Resolved(_entities)) => PluginTransition::None,
            (PluginState::Failed, payload) => {
                warn!("(failed) {:?}", &payload);

                PluginTransition::None
            }
            (state, message) => {
                warn!("(failing) {:?} {:?}", state, message);

                PluginTransition::Direct(PluginState::Failed)
            }
        }
    }
}

#[cfg(test)]
mod tests {
    #[allow(unused_imports)]
    use tracing::*;

    use crate::proto::{plugin::PluginState, Payload};

    use super::*;

    type TestPlugin = PluginProtocol<DefaultResponses>;

    #[tokio::test]
    async fn test_initialize() -> anyhow::Result<()> {
        let mut proto = TestPlugin::new_with_session_key("session-key".to_owned());

        assert_eq!(proto.machine.state, PluginState::Uninitialized);

        let session_key = proto.session_key().unwrap().to_owned();
        let initialize = Payload::Initialize(session_key);
        let mut sender = Default::default();
        proto.apply(&proto.message(initialize), &mut sender)?;

        assert_eq!(proto.machine.state, PluginState::Initialized);
        assert!(sender.queue.is_empty());

        Ok(())
    }
}
