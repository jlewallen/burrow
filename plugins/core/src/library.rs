pub mod plugin {
    pub use anyhow::Result;
    pub use serde::Deserialize;
    pub use tracing::*;

    pub use super::parser::try_parsing;
    pub use kernel::common::DeserializeTagged;
    pub use kernel::prelude::*;
    pub use kernel::try_deserialize_all;
}

pub mod parser {
    use nom::sequence::delimited;
    pub use nom::{
        branch::alt,
        bytes::complete::{tag, take_while, take_while1},
        character::complete::digit1,
        combinator::map,
        combinator::{map_res, opt, recognize},
        sequence::{pair, preceded, separated_pair, terminated, tuple},
        IResult,
    };
    pub use tracing::*;

    pub use kernel::prelude::*;

    pub fn word(i: &str) -> IResult<&str, &str> {
        take_while1(move |c| "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".contains(c))(i)
    }

    pub fn spaces(i: &str) -> IResult<&str, &str> {
        take_while1(move |c| " \t".contains(c))(i)
    }

    pub fn noun(i: &str) -> IResult<&str, Item> {
        map(word, |s: &str| Item::Named(s.to_owned()))(i)
    }

    pub fn text_to_end_of_line(i: &str) -> IResult<&str, &str> {
        take_while1(|_c: char| true)(i)
    }

    pub fn string_literal(i: &str) -> IResult<&str, &str> {
        delimited(tag("\""), string_inside, tag("\""))(i)
    }

    fn string_inside(i: &str) -> IResult<&str, &str> {
        take_while(|c: char| c.is_alphabetic() || c.is_whitespace())(i)
    }

    pub fn camel_case_word(i: &str) -> IResult<&str, &str> {
        word(i) // TODO
    }

    pub fn unsigned_number(i: &str) -> IResult<&str, u64> {
        map_res(recognize(digit1), str::parse)(i)
    }

    pub fn gid_reference(i: &str) -> IResult<&str, Item> {
        map(preceded(tag("#"), unsigned_number), |n| {
            Item::Gid(EntityGid::new(n))
        })(i)
    }

    pub fn surrounding_area(i: &str) -> IResult<&str, Item> {
        map(alt((tag("area"), tag("here"))), |_s: &str| Item::Area)(i)
    }

    pub fn noun_or_specific(i: &str) -> IResult<&str, Item> {
        alt((surrounding_area, myself, noun, gid_reference))(i)
    }

    pub fn myself(i: &str) -> IResult<&str, Item> {
        map(alt((tag("self"), tag("myself"))), |_s: &str| Item::Myself)(i)
    }

    pub fn named_place(i: &str) -> IResult<&str, Item> {
        alt((
            gid_reference,
            map(word, |s: &str| Item::Route(s.to_owned())),
        ))(i)
    }

    pub fn try_parsing<T: ParsesActions>(parser: T, i: &str) -> EvaluationResult {
        parser.try_parse_action(i)
    }
}

pub mod model {
    pub use anyhow::Result;
    pub use chrono::{DateTime, Utc};
    pub use kernel::common::*;
    pub use kernel::prelude::*;
    pub use macros::*;
    pub use serde::{Deserialize, Serialize};
    pub use serde_json::json;
    pub use std::rc::Rc;
    pub use std::{collections::HashMap, ops::Deref};
    pub use tracing::*;
}

pub mod actions {
    pub use crate::tools;
    pub use anyhow::Result;
    pub use kernel::common::*;
    pub use kernel::prelude::*;
    pub use macros::*;
    pub use serde::{Deserialize, Serialize};
    pub use tracing::*;

    pub fn reply_ok<T: ToTaggedJson + 'static>(
        living: EntityPtr,
        audience: Audience,
        raise: T,
    ) -> Result<Effect> {
        get_my_session()?.raise(
            Some(living),
            audience,
            Raising::TaggedJson(raise.to_tagged_json()?),
        )?;

        Ok(Effect::Ok)
    }

    pub fn reply_done<T: ToTaggedJson + 'static>(
        living: EntityPtr,
        audience: Audience,
        raise: T,
    ) -> Result<SimpleReply> {
        get_my_session()?.raise(
            Some(living),
            audience,
            Raising::TaggedJson(raise.to_tagged_json()?),
        )?;

        Ok(SimpleReply::Done)
    }
}

pub mod tests {
    pub use anyhow::Result;
    pub use chrono::TimeZone;
    pub use chrono::Utc;
    pub use serde::de::DeserializeOwned;
    pub use std::collections::HashSet;
    pub use std::rc::Rc;

    pub use crate::tools;
    pub use crate::{BuildSurroundings, QuickThing};
    pub use kernel::common::*;
    pub use kernel::here;
    pub use kernel::prelude::*;
    pub use tracing::*;

    pub use super::plugin::try_parsing;

    pub trait ToDebugJson {
        fn to_debug_json(&self) -> Result<JsonValue, serde_json::Error>;
    }

    impl ToDebugJson for Effect {
        fn to_debug_json(&self) -> Result<JsonValue, serde_json::Error> {
            serde_json::to_value(self)
        }
    }

    pub fn parse_and_perform<T: ParsesActions>(
        parser: T,
        line: &str,
    ) -> Result<(Surroundings, Effect)> {
        let mut build = BuildSurroundings::new()?;
        let (session, surroundings) = build.plain().encyclopedia()?.build()?;
        let action = try_parsing(parser, line)?;
        let action = action.unwrap();
        let effect = action.perform(session.clone(), &surroundings)?;
        build.close()?;
        Ok((surroundings, effect))
    }

    pub fn perform_directly<T: Action>(action: T) -> Result<(Surroundings, Effect)> {
        let mut build = BuildSurroundings::new()?;
        let (session, surroundings) = build.plain().encyclopedia()?.build()?;
        let action: Rc<dyn Action> = Rc::new(action);
        let effect = action.perform(session.clone(), &surroundings)?;
        build.close()?;
        Ok((surroundings, effect))
    }
}
