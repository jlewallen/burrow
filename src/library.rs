use nom::{bytes::complete::take_while1, combinator::map, IResult};

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Item {
    Described(String),
}

pub fn word(i: &str) -> IResult<&str, &str> {
    take_while1(move |c| "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".contains(c))(i)
}

pub fn spaces(i: &str) -> IResult<&str, &str> {
    take_while1(move |c| " \t".contains(c))(i)
}

pub fn noun(i: &str) -> IResult<&str, Item> {
    map(word, |s: &str| Item::Described(s.to_owned()))(i)
}
