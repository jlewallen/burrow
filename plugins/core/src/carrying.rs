use crate::library::plugin::*;

#[derive(Default)]
pub struct CarryingPluginFactory {}

impl PluginFactory for CarryingPluginFactory {
    fn create_plugin(&self) -> Result<Box<dyn Plugin>> {
        Ok(Box::new(CarryingPlugin {}))
    }

    fn stop(&self) -> Result<()> {
        Ok(())
    }
}

#[derive(Default)]
pub struct CarryingPlugin {}

impl Plugin for CarryingPlugin {
    fn plugin_key() -> &'static str
    where
        Self: Sized,
    {
        "carrying"
    }

    fn key(&self) -> &'static str {
        Self::plugin_key()
    }

    fn initialize(&mut self) -> Result<()> {
        Ok(())
    }

    fn middleware(&mut self) -> Result<Vec<Rc<dyn Middleware>>> {
        Ok(Vec::default())
    }

    fn deliver(&self, _incoming: &Incoming) -> Result<()> {
        Ok(())
    }

    fn stop(&self) -> Result<()> {
        Ok(())
    }
}

impl ParsesActions for CarryingPlugin {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        try_parsing(parser::DropActionParser {}, i)
            .or_else(|_| try_parsing(parser::HoldActionParser {}, i))
            .or_else(|_| try_parsing(parser::PutInsideActionParser {}, i))
            .or_else(|_| try_parsing(parser::TakeOutActionParser {}, i))
    }
}

pub mod model {
    use crate::{library::model::*, tools};

    pub type CarryingResult = Result<DomainOutcome>;

    #[derive(Debug, Serialize, ToJson)]
    #[serde(rename_all = "camelCase")]
    pub enum CarryingEvent {
        ItemHeld {
            living: EntityRef,
            item: EntityRef,
            area: EntityRef,
        },
        ItemDropped {
            living: EntityRef,
            item: EntityRef,
            area: EntityRef,
        },
    }

    impl DomainEvent for CarryingEvent {}

    #[derive(Debug, Serialize, Deserialize, Default)]
    pub struct Location {
        pub container: Option<EntityRef>,
    }

    impl Scope for Location {
        fn serialize(&self) -> Result<serde_json::Value> {
            Ok(serde_json::to_value(self)?)
        }

        fn scope_key() -> &'static str {
            "location"
        }
    }

    impl Needs<SessionRef> for Location {
        fn supply(&mut self, session: &SessionRef) -> Result<()> {
            self.container = session.ensure_optional_entity(&self.container)?;
            Ok(())
        }
    }

    #[derive(Debug, Serialize, Deserialize, Default)]
    pub struct Containing {
        pub holding: Vec<EntityRef>,
        pub capacity: Option<u32>,
        pub produces: HashMap<String, String>,
    }

    impl Scope for Containing {
        fn serialize(&self) -> Result<serde_json::Value> {
            Ok(serde_json::to_value(self)?)
        }

        fn scope_key() -> &'static str {
            "containing"
        }
    }

    impl Needs<SessionRef> for Containing {
        fn supply(&mut self, session: &SessionRef) -> Result<()> {
            self.holding = self
                .holding
                .iter()
                .map(|r| session.ensure_entity(r))
                .collect::<Result<Vec<_>, DomainError>>()?;

            Ok(())
        }
    }

    impl Containing {
        pub fn start_carrying(&mut self, item: &Entry) -> CarryingResult {
            let carryable = item.scope::<Carryable>()?;

            let holding = self
                .holding
                .iter()
                .map(|h| h.to_entry())
                .collect::<Result<Vec<_>, _>>()?;

            for held in holding {
                if is_kind(&held, &carryable.kind)? {
                    let mut combining = held.scope_mut::<Carryable>()?;

                    combining.increase_quantity(carryable.quantity)?;

                    combining.save()?;

                    get_my_session()?.obliterate(item)?;

                    return Ok(DomainOutcome::Ok);
                }
            }

            self.holding.push(item.try_into()?);

            Ok(DomainOutcome::Ok)
        }

        pub fn is_holding(&self, item: &Entry) -> Result<bool> {
            Ok(self.holding.iter().any(|i| *i.key() == *item.key()))
        }

        fn remove_item(&mut self, item: &Entry) -> CarryingResult {
            self.holding = self
                .holding
                .iter()
                .flat_map(|i| {
                    if *i.key() == *item.key() {
                        vec![]
                    } else {
                        vec![i.clone()]
                    }
                })
                .collect::<Vec<EntityRef>>()
                .to_vec();

            Ok(DomainOutcome::Ok)
        }

        pub fn stop_carrying(&mut self, item: &Entry) -> Result<Option<Entry>> {
            if !self.is_holding(item)? {
                return Ok(None);
            }

            let carryable = item.scope_mut::<Carryable>()?;
            if carryable.quantity > 1.0 {
                let (_original, separated) = tools::separate(item, 1.0)?;

                Ok(Some(separated))
            } else {
                self.remove_item(item)?;

                Ok(Some(item.clone()))
            }
        }
    }

    #[derive(Debug, Serialize, Deserialize)]
    pub struct Carryable {
        kind: Kind,
        quantity: f32,
    }

    fn is_kind(entity: &Entry, kind: &Kind) -> Result<bool> {
        Ok(*entity.scope::<Carryable>()?.kind() == *kind)
    }

    impl Default for Carryable {
        fn default() -> Self {
            let session = get_my_session().expect("No session in Entity::new_blank!");
            Self {
                kind: Kind::new(session.new_identity()),
                quantity: 1.0,
            }
        }
    }

    impl Carryable {
        pub fn quantity(&self) -> f32 {
            self.quantity
        }

        pub fn decrease_quantity(&mut self, q: f32) -> Result<&mut Self, DomainError> {
            self.sanity_check_quantity();

            if q < 1.0 || q > self.quantity {
                Err(DomainError::Impossible)
            } else {
                self.quantity -= q;

                Ok(self)
            }
        }

        pub fn increase_quantity(&mut self, q: f32) -> Result<&mut Self> {
            self.sanity_check_quantity();

            self.quantity += q;

            Ok(self)
        }

        pub fn set_quantity(&mut self, q: f32) -> Result<&mut Self> {
            self.quantity = q;

            Ok(self)
        }

        pub fn kind(&self) -> &Kind {
            &self.kind
        }

        pub fn set_kind(&mut self, kind: &Kind) {
            self.kind = kind.clone();
        }

        // Migrate items that were initialized with 0 quantities.
        fn sanity_check_quantity(&mut self) {
            if self.quantity < 1.0 {
                self.quantity = 1.0
            }
        }
    }

    impl Scope for Carryable {
        fn serialize(&self) -> Result<serde_json::Value> {
            Ok(serde_json::to_value(self)?)
        }

        fn scope_key() -> &'static str {
            "carryable"
        }
    }

    impl Needs<SessionRef> for Carryable {
        fn supply(&mut self, _session: &SessionRef) -> Result<()> {
            Ok(())
        }
    }
}

pub mod actions {
    use crate::{carrying::model::CarryingEvent, library::actions::*};

    #[action]
    pub struct HoldAction {
        pub item: Item,
    }

    impl Action for HoldAction {
        fn is_read_only() -> bool {
            false
        }

        fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
            info!("hold {:?}!", self.item);

            let (_, user, area) = surroundings.unpack();

            match session.find_item(surroundings, &self.item)? {
                Some(holding) => match tools::move_between(&area, &user, &holding)? {
                    DomainOutcome::Ok => Ok(reply_done(
                        Audience::Area(area.key().clone()),
                        CarryingEvent::ItemHeld {
                            living: user.entity_ref(),
                            item: holding.entity_ref(),
                            area: area.entity_ref(),
                        },
                    )?
                    .into()),
                    DomainOutcome::Nope => Ok(SimpleReply::NotFound.into()),
                },
                None => Ok(SimpleReply::NotFound.into()),
            }
        }
    }

    #[action]
    pub struct DropAction {
        pub maybe_item: Option<Item>,
    }

    impl Action for DropAction {
        fn is_read_only() -> bool {
            false
        }

        fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
            info!("drop {:?}!", self.maybe_item);

            let (_, user, area) = surroundings.unpack();

            match &self.maybe_item {
                Some(item) => match session.find_item(surroundings, item)? {
                    Some(dropping) => match tools::move_between(&user, &area, &dropping)? {
                        DomainOutcome::Ok => Ok(reply_done(
                            Audience::Area(area.key().clone()),
                            CarryingEvent::ItemDropped {
                                living: user.entity_ref(),
                                item: dropping.entity_ref(),
                                area: area.entity_ref(),
                            },
                        )?
                        .into()),
                        DomainOutcome::Nope => Ok(SimpleReply::NotFound.into()),
                    },
                    None => Ok(SimpleReply::NotFound.into()),
                },
                None => Ok(SimpleReply::NotFound.into()),
            }
        }
    }

    #[action]
    pub struct PutInsideAction {
        pub item: Item,
        pub vessel: Item,
    }

    impl Action for PutInsideAction {
        fn is_read_only() -> bool {
            false
        }

        fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
            info!("put-inside {:?} -> {:?}", self.item, self.vessel);

            let (_, _user, _area) = surroundings.unpack();

            match session.find_item(surroundings, &self.item)? {
                Some(item) => match session.find_item(surroundings, &self.vessel)? {
                    Some(vessel) => {
                        if tools::is_container(&vessel)? {
                            let from = tools::container_of(&item)?;
                            match tools::move_between(&from.try_into()?, &vessel, &item)? {
                                DomainOutcome::Ok => Ok(SimpleReply::Done.into()),
                                DomainOutcome::Nope => Ok(SimpleReply::NotFound.into()),
                            }
                        } else {
                            Ok(SimpleReply::Impossible.into())
                        }
                    }
                    None => Ok(SimpleReply::NotFound.into()),
                },
                None => Ok(SimpleReply::NotFound.into()),
            }
        }
    }

    #[action]
    pub struct TakeOutAction {
        pub item: Item,
        pub vessel: Item,
    }

    impl Action for TakeOutAction {
        fn is_read_only() -> bool {
            false
        }

        fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
            info!("take-out {:?} -> {:?}", self.item, self.vessel);

            let (_, user, _area) = surroundings.unpack();

            match session.find_item(surroundings, &self.vessel)? {
                Some(vessel) => {
                    if tools::is_container(&vessel)? {
                        match session.find_item(surroundings, &self.item)? {
                            Some(item) => match tools::move_between(&vessel, &user, &item)? {
                                DomainOutcome::Ok => Ok(SimpleReply::Done.into()),
                                DomainOutcome::Nope => Ok(SimpleReply::NotFound.into()),
                            },
                            None => Ok(SimpleReply::NotFound.into()),
                        }
                    } else {
                        Ok(SimpleReply::Impossible.into())
                    }
                }
                None => Ok(SimpleReply::NotFound.into()),
            }
        }
    }
}

pub mod parser {
    use super::actions::*;
    use crate::library::parser::*;

    pub struct HoldActionParser {}

    impl ParsesActions for HoldActionParser {
        fn try_parse_action(&self, i: &str) -> EvaluationResult {
            let (_, action) = map(separated_pair(tag("hold"), spaces, noun), |(_, target)| {
                HoldAction { item: target }
            })(i)?;

            Ok(Some(Box::new(action)))
        }
    }

    pub struct DropActionParser {}

    impl ParsesActions for DropActionParser {
        fn try_parse_action(&self, i: &str) -> EvaluationResult {
            let specific = map(separated_pair(tag("drop"), spaces, noun), |(_, target)| {
                DropAction {
                    maybe_item: Some(Item::Held(Box::new(target))),
                }
            });

            let everything = map(tag("drop"), |_| DropAction { maybe_item: None });

            let (_, action) = alt((specific, everything))(i)?;

            Ok(Some(Box::new(action)))
        }
    }

    pub struct TakeOutActionParser {}

    impl ParsesActions for TakeOutActionParser {
        fn try_parse_action(&self, i: &str) -> EvaluationResult {
            let item = map(separated_pair(tag("take"), spaces, noun), |(_, target)| {
                target
            });

            let (_, action) = map(
                separated_pair(separated_pair(item, spaces, tag("out of")), spaces, noun),
                |(item, target)| TakeOutAction {
                    item: Item::Contained(Box::new(item.0)),
                    vessel: target,
                },
            )(i)?;

            Ok(Some(Box::new(action)))
        }
    }

    pub struct PutInsideActionParser {}

    impl ParsesActions for PutInsideActionParser {
        fn try_parse_action(&self, i: &str) -> EvaluationResult {
            let item = map(separated_pair(tag("put"), spaces, noun), |(_, target)| {
                target
            });

            let (_, action) = map(
                separated_pair(
                    separated_pair(
                        item,
                        spaces,
                        pair(tag("inside"), opt(pair(spaces, tag("of")))),
                    ),
                    spaces,
                    noun,
                ),
                |(item, target)| PutInsideAction {
                    item: item.0,
                    vessel: target,
                },
            )(i)?;

            Ok(Some(Box::new(action)))
        }
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashSet;

    use super::parser::*;
    use super::*;
    use crate::carrying::model::Location;
    use crate::tools;
    use crate::{
        carrying::model::Containing,
        {BuildSurroundings, QuickThing},
    };

    #[test]
    fn it_holds_unheld_items() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let (session, surroundings) = build
            .ground(vec![QuickThing::Object("Cool Rake")])
            .build()?;

        let (_, person, area) = surroundings.unpack();
        assert_eq!(person.scope::<Containing>()?.holding.len(), 0);
        assert_eq!(area.scope::<Containing>()?.holding.len(), 1);

        let action = try_parsing(HoldActionParser {}, "hold rake")?;
        let action = action.unwrap();
        let reply = action.perform(session.clone(), &surroundings)?;

        assert_eq!(reply.to_tagged_json()?, SimpleReply::Done.to_tagged_json()?);

        assert_eq!(person.scope::<Containing>()?.holding.len(), 1);
        assert_eq!(area.scope::<Containing>()?.holding.len(), 0);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_separates_multiple_ground_items_when_held() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let (session, surroundings) = build
            .ground(vec![QuickThing::Multiple("Cool Rake", 2.0)])
            .build()?;

        let (_, person, area) = surroundings.unpack();
        assert_eq!(person.scope::<Containing>()?.holding.len(), 0);
        assert_eq!(area.scope::<Containing>()?.holding.len(), 1);

        let action = try_parsing(HoldActionParser {}, "hold rake")?;
        let action = action.unwrap();
        let reply = action.perform(session.clone(), &surroundings)?;

        assert_eq!(reply.to_tagged_json()?, SimpleReply::Done.to_tagged_json()?);

        let held = &person.scope::<Containing>()?.holding;
        let ground = &area.scope::<Containing>()?.holding;
        assert_eq!(held.len(), 1);
        assert_eq!(ground.len(), 1);

        let held_keys: HashSet<_> = held.iter().map(|i| i.key().clone()).collect();
        let ground_keys: HashSet<_> = ground.iter().map(|i| i.key().clone()).collect();
        let common_keys: HashSet<_> = held_keys.intersection(&ground_keys).collect();
        assert_eq!(common_keys.len(), 0);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_combines_multiple_items_when_together_on_ground() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let same_kind = build.make(QuickThing::Object("Cool Rake"))?;
        tools::set_quantity(&same_kind, 2.0)?;
        let (first, second) = tools::separate(&same_kind, 1.0)?;
        let (session, surroundings) = build
            .ground(vec![QuickThing::Actual(first.clone())])
            .hands(vec![QuickThing::Actual(second)])
            .build()?;

        let (_, person, area) = surroundings.unpack();
        assert_eq!(person.scope::<Containing>()?.holding.len(), 1);
        assert_eq!(area.scope::<Containing>()?.holding.len(), 1);

        let action = try_parsing(HoldActionParser {}, "hold rake")?;
        let action = action.unwrap();
        let reply = action.perform(session.clone(), &surroundings)?;

        assert_eq!(reply.to_tagged_json()?, SimpleReply::Done.to_tagged_json()?);

        assert_eq!(person.scope::<Containing>()?.holding.len(), 1);
        assert_eq!(area.scope::<Containing>()?.holding.len(), 0);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_fails_to_hold_unknown_items() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let (session, surroundings) = build
            .ground(vec![QuickThing::Object("Cool Broom")])
            .build()?;

        let action = try_parsing(HoldActionParser {}, "hold rake")?;
        let action = action.unwrap();
        let reply = action.perform(session.clone(), &surroundings)?;
        let (_, person, area) = surroundings.unpack();

        assert_eq!(
            reply.to_tagged_json()?,
            SimpleReply::NotFound.to_tagged_json()?
        );

        assert_eq!(person.scope::<Containing>()?.holding.len(), 0);
        assert_eq!(area.scope::<Containing>()?.holding.len(), 1);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_drops_held_items() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let (session, surroundings) = build.hands(vec![QuickThing::Object("Cool Rake")]).build()?;

        let action = try_parsing(DropActionParser {}, "drop rake")?;
        let action = action.unwrap();
        let reply = action.perform(session.clone(), &surroundings)?;
        let (_, person, area) = surroundings.unpack();

        assert_eq!(reply.to_tagged_json()?, SimpleReply::Done.to_tagged_json()?);

        assert_eq!(person.scope::<Containing>()?.holding.len(), 0);
        assert_eq!(area.scope::<Containing>()?.holding.len(), 1);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_fails_to_drop_unknown_items() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let (session, surroundings) = build
            .hands(vec![QuickThing::Object("Cool Broom")])
            .build()?;

        let action = try_parsing(DropActionParser {}, "drop rake")?;
        let action = action.unwrap();
        let reply = action.perform(session.clone(), &surroundings)?;
        let (_, person, area) = surroundings.unpack();

        assert_eq!(
            reply.to_tagged_json()?,
            SimpleReply::NotFound.to_tagged_json()?
        );

        assert_eq!(person.scope::<Containing>()?.holding.len(), 1);
        assert_eq!(area.scope::<Containing>()?.holding.len(), 0);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_fails_to_drop_unheld_items() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let (session, surroundings) = build
            .ground(vec![QuickThing::Object("Cool Broom")])
            .build()?;

        let action = try_parsing(DropActionParser {}, "drop rake")?;
        let action = action.unwrap();
        let reply = action.perform(session.clone(), &surroundings)?;
        let (_, person, area) = surroundings.unpack();

        assert_eq!(
            reply.to_tagged_json()?,
            SimpleReply::NotFound.to_tagged_json()?
        );

        assert_eq!(person.scope::<Containing>()?.holding.len(), 0);
        assert_eq!(area.scope::<Containing>()?.holding.len(), 1);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_fails_to_puts_item_in_non_containers() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let vessel = build.entity()?.named("Not A Vessel")?.into_entry()?;
        let (session, surroundings) = build
            .hands(vec![
                QuickThing::Object("key"),
                QuickThing::Actual(vessel.clone()),
            ])
            .build()?;

        let action = try_parsing(PutInsideActionParser {}, "put key inside vessel")?;
        let action = action.unwrap();
        let reply = action.perform(session.clone(), &surroundings)?;
        let (_world, person, _area) = surroundings.unpack();

        insta::assert_json_snapshot!(reply.to_tagged_json()?);

        assert_eq!(person.scope::<Containing>()?.holding.len(), 2);
        assert_eq!(vessel.scope::<Containing>()?.holding.len(), 0);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_puts_items_in_containers() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let vessel = build
            .entity()?
            .named("Vessel")?
            .holding(&vec![])?
            .into_entry()?;
        let (session, surroundings) = build
            .hands(vec![
                QuickThing::Object("key"),
                QuickThing::Actual(vessel.clone()),
            ])
            .build()?;

        let action = try_parsing(PutInsideActionParser {}, "put key inside vessel")?;
        let action = action.unwrap();
        let reply = action.perform(session.clone(), &surroundings)?;
        let (_world, person, _area) = surroundings.unpack();

        insta::assert_json_snapshot!(reply.to_tagged_json()?);

        assert_eq!(person.scope::<Containing>()?.holding.len(), 1);
        assert_eq!(vessel.scope::<Containing>()?.holding.len(), 1);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_takes_items_out_of_containers() -> Result<()> {
        let mut build = BuildSurroundings::new()?;
        let key = build.entity()?.named("Key")?.into_entry()?;
        let vessel = build
            .entity()?
            .named("Vessel")?
            .holding(&vec![key.clone()])?
            .into_entry()?;
        let (session, surroundings) = build
            .hands(vec![QuickThing::Actual(vessel.clone())])
            .build()?;

        let action = try_parsing(TakeOutActionParser {}, "take key out of vessel")?;
        let action = action.unwrap();
        let reply = action.perform(session.clone(), &surroundings)?;
        let (_world, person, _area) = surroundings.unpack();

        insta::assert_json_snapshot!(reply.to_tagged_json()?);

        assert_eq!(person.scope::<Containing>()?.holding.len(), 2);
        assert_eq!(vessel.scope::<Containing>()?.holding.len(), 0);
        assert_eq!(
            *key.scope::<Location>()?.container.as_ref().unwrap().key(),
            *person.key()
        );

        build.close()?;

        Ok(())
    }
}
