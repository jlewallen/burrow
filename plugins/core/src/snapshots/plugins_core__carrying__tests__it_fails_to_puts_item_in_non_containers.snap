---
source: plugins/core/src/carrying.rs
expression: reply.to_json()?
---
{
  "simple": "impossible"
}
